Z: remove unused and redundant register from object evaluator

ehsankianifar opened the pull request:
Repurpose registers when no longer needed.
Remove unused registers.
Use scratch register manager inside genHeapAlloc

ehsankianifar commented:
@r30shah please review when possible. Thanks.
Sanity tests pass with disabled batch clear (build 23202)

ehsankianifar commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10202,24 +10205,8 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
-            }
-
-         iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BNE, node, callLabel, iCursor);
+         iCursor = generateRILInstruction(cg, TR::InstOpCode::CLFI, node, enumReg, (1 << 16), iCursor);
```
To eliminate the need for a register (tmp)

r30shah commented:
Hi @ehsankianifar  please update the commit body with appropriate explanation of what the change is about and shorten the commit title withing 72 letters.

r30shah commented:
@ehsankianifar  The build you launched was only ran on x86, not LoZ (https://github.com/eclipse-openj9/openj9/pull/19860#issuecomment-2229906284). 

I am still suspicious of some of the registers you removed as there are many different type of cases this will be used so I am going to do thorough review today. 

Given the scope of area this change will be touching, please launch internal jenkins build on Linux on Z  with JDK11 and JDK21 for sanity and extended test target (functional / system and openjdk). Also launch Java 8 vmfarm build on z/OS and Linux on Z with c,p target as well.

ehsankianifar commented:
Thanks @r30shah for checking that. I start a new build right now

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10730,16 +10675,11 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    TR::Register *enumReg               = NULL;
    TR::Register *copyReg               = NULL;
-   TR::Register *classRegAOT           = NULL;
    TR::Register *temp1Reg              = NULL;
```
As you are using temp1Reg to hold the data size of the array / object, can we rename it to make code more readable ?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10895,15 +10826,12 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
-      // if variable-length array   - dataSizeReg will contain the (calculated) size
-      // if outlined                - tmpReg will contain the value of
+      // Two possible outputs:
+      // if variable-length array   - temp1Reg will contain the (calculated) size
```
As suggested previously, rename temp1Reg to something more meaningful and change the comment here as well.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10941,24 +10869,7 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
 
```
Are comments from line 100849-100866 relevant ?

ehsankianifar replied:
It does not match the current instruction selection. I remove it

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10794,10 +10731,6 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
-         }
       TR::Node * firstChild = node->getFirstChild();
       TR::Node * secondChild = NULL;
 
```
While you are here, can you also tabs at line 10762-10772 as well?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11041,14 +10952,14 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
                {
                if (comp->target().is64Bit())
-               iCursor = generateRSInstruction(cg, TR::InstOpCode::SLLG, node, tmp, tmp, trailingZeroes(elementSize), iCursor);
+               iCursor = generateRSInstruction(cg, TR::InstOpCode::SLLG, node, temp1Reg, temp1Reg, trailingZeroes(elementSize), iCursor);
```
Please fix tabs here and at line 10962 as well

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11131,11 +11041,11 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
          if ( comp->compileRelocatableCode() && opCode == TR::anewarray)
-         genInitArrayHeader(node, iCursor, isVariableLen, classAddress, classReg, resReg, zeroReg,
-               enumReg, dataSizeReg, temp1Reg, litPoolBaseReg, conditions, cg);
+         genInitArrayHeader(node, iCursor, isVariableLen, classAddress, classReg, resReg,
```
Seems like for anewarray and relocatable code, only parameter that is passed is classReg, can we shrink this to one statement with conditional assignment ?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11233,7 +11143,7 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
       else
          {
-         genInitObjectHeader(node, iCursor, classAddress, classReg , resReg, zeroReg, temp1Reg, litPoolBaseReg, conditions, cg);
+         genInitObjectHeader(node, iCursor, classAddress, classReg , resReg, temp1Reg, litPoolBaseReg, conditions, cg);
```
Please space the arguments consistently.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10271,25 +10249,24 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
+      // Zero initialize newly allocated array or object if batch clearing is disabled and node can not skip zero initialization.
+      static bool disableBatchClear = feGetEnv("TR_DisableBatchClear") != NULL;
+      bool inlineZeroInitialization = disableBatchClear && !node->canSkipZeroInitialization();
+      TR::Register *sizePreserveRegister = NULL;
```
sizePreserveRegister seems bit out of context (If I have recommended in last PR to rename it, I apologize), but lengthReg seems to be fine. No need to rename it.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10363,82 +10340,66 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
+            TR::Register * addressReg = srm->findOrCreateScratchRegister();
+            iCursor = generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, addressReg, resReg, iCursor);
+
+            TR::LabelSymbol * ResidueBytesZeroingLabel = generateLabelSymbol(cg);
```
variable names are lowerCamelCase throughout the code-base. Please don't diverge from that.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10363,82 +10340,66 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
             }
          else
             {
+            int32_t loops = allocSize / 256;
```
May be this may have escaped previous PR, but loops sounds like number of loop you will be generating. May be `loopIterCount` or something synonymous can be used here?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10363,82 +10340,66 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
-
+               // Repurpose sizeReg to keep address.
+               TR::Register * addressReg = sizeReg;
+               TR::Register * loopCountRegister = srm->findOrCreateScratchRegister();
```
Same as previous suggestion, loopIterRegsiter may be a better name here.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10363,82 +10340,66 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
-               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
-               iCursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, addressReg, generateS390MemoryReference(addressReg, 256, cg), iCursor);
-               iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, shiftReg, fillerLoopLabel);
+               iCursor = generateLoad32BitConstant(cg, node, loops, loopCountRegister, true, iCursor, conditions);
```
When you load the loopIterCount to a register, we do not need to set the condition code (`true` parameter passed in for generateLoad32BitConstant), can you fix that?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10176,10 +10170,19 @@ roundArrayLengthToObjectAlignment(TR::CodeGenerator* cg, TR::Node* node, TR::Ins
+static void
+genMemoryZeroingLoop(TR::CodeGenerator* cg, TR::Node* node, TR::Instruction*& iCursor, TR::Register * addressReg, TR::Register * loopCountRegister)
+   {
+   TR::LabelSymbol * loopTopLabel = generateLabelSymbol(cg);
```
Instead of `loopTopLabel`, I thin, startLabel/loopStartLabel would be more meaningful here.

r30shah reviewed (changes requested):
Thanks @ehsankianifar for going in the evaluator and cleaning up some part of the code. Overall looked OK to me,  but I would like to give one more look specially around the use of Scratch register manager. I have requested some minor changes (Mostly clarifying comments and renaming some of the variable names).  Please make this changes as seperate single commit so it is easier to review.

ehsankianifar commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10156,12 +10156,6 @@ roundArrayLengthToObjectAlignment(TR::CodeGenerator* cg, TR::Node* node, TR::Ins
    bool needsAlignment = ( ((allocSize % alignmentConstant) != 0) ||
                            ((elementSize % alignmentConstant) != 0) );
 
-   bool canCombineAGRs = ( ((allocSize % alignmentConstant) == 0) &&
```
inside VMnewEvaluator (currently line 10944) the code checks this condition for canCombineAGRs and if it is true, save the size in temp1Reg (which is sizeReg here) and if the condition is false, save the size in dataSizeReg.
the value of those registers never change from there (VMnewEvaluatior line ~ 10944) to here.
It is unnecessary to save the size in a dataSizeReg and load it to sizeReg here. I removed the condition from both places and used only sizeReg (temp1Reg).

ehsankianifar commented:
@r30shah I addressed the comments in a separate commit as you requested. I can squash the commits after the review or leave it as it is. The tests are passing for the java 8 with c,p tests target on zos and zLinux (build number 34097).  
The sanity and extended tests mostly passed with java 11 and 21 on zLinux (build 23269). 
some tests are still running. I will check later and post update.

ehsankianifar commented:
sanity.functional tests were aborted on both jdk21 and jdk11 builds. both passed on rerun (grinder 42320 and 42321)
all other tests passed as well.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10678,7 +10677,7 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    TR::Register *enumReg               = NULL;
    TR::Register *copyReg               = NULL;
-   TR::Register *temp1Reg              = NULL;
+   TR::Register *dataSizeReg              = NULL;
```
Please align NULL with other lines.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11130,12 +11010,8 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
-         else
-         genInitArrayHeader(node, iCursor, isVariableLen, classAddress, NULL, resReg, zeroReg,
-               enumReg, dataSizeReg, temp1Reg, litPoolBaseReg, conditions, cg);
+         genInitArrayHeader(node, iCursor, isVariableLen, classAddress, (comp->compileRelocatableCode() && opCode == TR::anewarray) ? classReg : NULL,
```
```suggestion
            genInitArrayHeader(node, iCursor, isVariableLen, classAddress,
                                    (comp->compileRelocatableCode() && opCode == TR::anewarray) ? classReg : NULL,
                                    resReg, enumReg, dataSizeReg, ...
```

r30shah reviewed (approved):
Small nitpicks, please squash the commits and move to from draft to ready for review.

ehsankianifar commented:
@r30shah commits squashed!

r30shah commented:
jenkins test sanity zlinux jdk21
